src/avt/Database/Database/avtGenericDatabase.C

```diff
@@ -5146,6 +5150,14 @@ avtGenericDatabase::MaterialSelect(vtkDataSet *ds, avtMaterial *mat,
         return NULL;
     }
     int topoDim = mmd->topologicalDimension;
+    if (ds->GetDataObjectType() == VTK_STRUCTURED_GRID)
+    {
+        int *dims = ((vtkStructuredGrid*)ds)->GetDimensions();
+        if ((dims[0] == 1 && dims[1] > 1 && dims[2] > 1) ||
+            (dims[1] == 1 && dims[0] > 1 && dims[2] > 1) ||
+            (dims[2] == 1 && dims[0] > 1 && dims[1] > 1))
+            topoDim = 2;
+    }
```

Comment:
I was suggesting Move the logic. The topological dimension calculation is being done in avtSourceFromDatabase::FetchDataset, which is performed before MaterialSelect, correct?
The only catch I see is that you are retrieving Topo dim from avtDatabaseMetaData, but the calculated topo dim is stored in the pipeline's avtDataAttributes.

Reply:
I'll take a look 👀

Reply:
First, this line, https://github.com/visit-dav/visit/blob/7092751482520c1e3a4c6b50d7fe1722791b429f/src/avt/Database/Database/avtGenericDatabase.C#L5152

was already there. That is what was originally setting topo dim from `avtDatabaseMetaData`, I just override it with a quick check of dimensions for a structured grid from the `vtkDataset` object itself. Next, I can't confirm order of FetchDataset and MaterialSelect here. I mean, it sounds correct that we'd have to fetch a dataset before we can material select it. Finally, I don't see any `avtDataAttributes` *nearby* here. In fact, there are no `avtDataAttributes` anywhere in `avtGenericDatabase`.

Reply:
It looks like, to get the calculated topoDim into this MaterialSelect method, would have to retrieve it via `src->GetOutput()->GetInfo().GetAttributes().GetTopologicalDimension()` from the other MaterialSelect, and add it as an argument in its call to [MaterialSelect](https://github.com/visit-dav/visit/blob/7092751482520c1e3a4c6b50d7fe1722791b429f/src/avt/Database/Database/avtGenericDatabase.C#L9672)

Reply:
@biagas question here...we have a dataset *collection* here...several `vtkDataset` objects. I worry that each `vtkDataset` might have a different answer to what its topological dimension is (and as things are currently coded, this is accomodated fine) and that if we take a single value for all members of the collection, we create problems.

Thoughts?

Reply:
In other words, won't the new `topoDim` argument we're considering adding to pass into `MaterialSelect` be the same across all datasets irrespective of the individual dataset's shapes?

I think we had decided that `src->GetOutput()->GetInfo().GetAttributes().GetTopologicalDimension()` is really a *maximum* topological dimension, right?

Reply:
> Will not having an accurate per-vtkDataSet topological dimension cause problems for other mesh types in MaterialSelect?

I think this is surely possible...maybe even likely. I think we potentially have a systemic issue throughout VisIt with how topological dimension was historically assumed to be equivalent to spatial dimension. Over time, that assumption has been tackled in various corner cases. But, I don't think its been handled fully generally and properly everywhere.

What I think about this PR relative to *that* question is that this PR makes this matter no *worse* than it was before. It improves it slightly for 2D structured gridded surfaces in 3 space.

All that said, if it helps, I would be happy to add the *equivalent* logic to `avtCommonDataFunctions.C:CGetTopologicalDim()` so that function at least reports a similar output when asked under the same circumstances.

Reply:
> What I think about this PR relative to _that_ question is that this PR makes this matter no _worse_ than it was before. It improves it slightly for 2D structured gridded surfaces in 3 space.

I agree.  I don't think you need to add anything else to satisfy the scope of work for this PR. Thank you for considering my suggestions.

Reply:
@biagas have a look at https://github.com/visit-dav/visit/blob/daa5e86ca8e2e69c473a37ab3aad3439424571f1/src/avt/Pipeline/Data/avtCommonDataFunctions.C#L3811-L3850

I added logic for returning valid value for structured gridded surfaces (and curves) in 3 space.